### PR TITLE
fix: fall back to hashed signing when Ledger rejects EIP-712 clear signing

### PR DESCRIPTION
**Pull request by ivan-angjelkoski**
### The problem

When a user signs an `eth_signTypedData_v4` message through MetaMask while a Ledger hardware wallet is the active keyring, signing fails for large payloads. The UI surfaces **"Blind signing is not enabled"** even when blind signing is already enabled on the device — the dapp receives no usable signature and the user has no recovery path.

Reproducing payloads seen in the wild:
- Typed data containing large stringified contract-call JSON
- Multi-message Cosmos-SDK EIP-712 payloads
- EIP-712 messages with deeply nested or numerous domain types

### Root cause

1. The dapp request reaches `signTypedData` in [`app/offscreen/hardware-wallets/ledger.ts`](https://github.com/MetaMask/metamask-extension/blob/main/app/offscreen/hardware-wallets/ledger.ts), which calls `app.signEIP712Message(hdPath, message)`.
2. For oversized or unparseable payloads, the Ledger Ethereum app's filtering/resolver cannot handle the data and the device responds with a status code such as **`0x6808`** (`CONDITIONS_OF_USE_NOT_SATISFIED`) — **not** `INS_NOT_SUPPORTED`.
3. The existing fallback helper `#isInsNotSupported` only matched `statusText === 'INS_NOT_SUPPORTED'`, so `0x6808` was re-thrown without retrying via `signEIP712HashedMessage`.
4. The thrown error is mapped by `LEDGER_ERROR_MAPPINGS` in [`ui/contexts/hardware-wallets/rpcErrorUtils.ts`](https://github.com/MetaMask/metamask-extension/blob/main/ui/contexts/hardware-wallets/rpcErrorUtils.ts) to `ErrorCode.DeviceStateBlindSignNotSupported`, which renders as **"Blind signing is not enabled"** — a message intended for users who have not toggled the device setting on, but which now fires even when they have.

The existing `INS_NOT_SUPPORTED` → hashed-signing fallback (added in #41017) correctly handles older devices (Nano S). It does not handle newer firmware (Nano X / S Plus / Stax) that rejects oversized payloads with a different status code.

A secondary failure mode has also been observed: some firmware versions silently sign a truncated representation of an oversized payload with **no error at all**. The returned signature then fails downstream verification with "The signature doesn't match the right address".

### The fix

Two complementary changes in `app/offscreen/hardware-wallets/ledger.ts`, both behind the same `signEIP712Message` → `signEIP712HashedMessage` fallback pattern that already exists:

1. **Widen the error-triggered fallback** from "only `INS_NOT_SUPPORTED`" to "any error whose status code indicates the device could not clear-sign". Covers `0x6808`, `0x6a80`, `0x6985`, plus the existing `INS_NOT_SUPPORTED` path. Helper renamed `#isInsNotSupported` → `#shouldFallBackToHashedSigning`. This resolves the reported bug.
2. **Add verify-after-sign as defense in depth** — recover the signer from any returned clear-sign signature via `recoverTypedSignature`; if it does not match the HD-path address, retry via hashed signing. No impact on the happy path (no extra device prompts — the check is a local `ecrecover`); protects against the silent-truncation firmware bug. New helper `#verifyTypedDataSignature`.

If the hashed-signing retry also fails verification, a `HardwareWalletError` is thrown with a clear message rather than returning an invalid signature.

### Why not…

| Approach | Verdict |
| --- | --- |
| Match on the UI error string ("Blind signing is not enabled") upstream | Rejected — that string is synthesised by `LEDGER_ERROR_MAPPINGS` in the UI layer, far downstream from the offscreen signer. Plumbing the error back is fragile and couples modules. |
| Pre-check payload size and always use hashed when large | Rejected — encodes device-specific thresholds, degrades UX for small payloads, doesn't help when the device itself is the source of truth for "can I clear-sign this?". |
| Fall back on ANY clear-sign error | Rejected — would mask unrelated failures (device locked, user rejected, transport errors) by retrying a sign that should surface the original error to the user. |
| Widen error matching to known clear-sign rejection codes + verify-after-sign | ✅ Resolves the reported bug (error-path) plus future-proofs against silent-truncation (no-error path). No UX regression on the happy path. |

### User experience

- **Happy path (any payload that clear-signs correctly)**: one on-device prompt, unchanged. The verify step runs locally after the device returns the signature and is invisible to the user.
- **Bug path (oversized payload, device rejects with `0x6808`)**: one prompt — device rejects, MetaMask silently retries via hashed signing, device shows the second prompt with hashes, user confirms once. Same total prompts as today's `INS_NOT_SUPPORTED` path on Nano S.
- **Silent-truncation path (device returns a wrong signature without error)**: two prompts — first on the clear-sign attempt (produces wrong sig), second on the hashed-sign retry. Two prompts is the minimum possible recovery from that firmware state; the alternative is today's broken behaviour where signing fails with no recovery.

## **Changelog**

CHANGELOG entry: Fixed EIP-712 typed data signing on Ledger devices for large payloads, which previously failed with a misleading "Blind signing is not enabled" error even when blind signing was enabled; the signer now retries via hashed signing when clear signing is rejected, and verifies the returned signature against the expected address.

## **Related issues**

Builds on #41017 (prior fix that introduced the `INS_NOT_SUPPORTED` fallback for Nano S).

## **Manual testing steps**

Unit tests cover all branches; an end-to-end repro requires physical Ledger hardware.

1. **Unit tests** — `yarn jest app/offscreen/hardware-wallets/ledger.test.ts`. All cases pass (3 updated + 3 new).
2. **Type check** — `yarn tsc --noEmit` confirms the updated helper/return types.
3. **Local build** — `yarn start`, load unpacked into Chrome.
4. **Repro payload** — connect a Ledger Nano X / S Plus / Stax on the latest Ethereum app with blind signing **enabled**; trigger `eth_signTypedData_v4` whose payload exceeds the device's clear-signing buffer.
   - Before fix: fails with UI message "Blind signing is not enabled".
   - After fix: device prompts with hashes, user confirms, signature broadcast succeeds.
5. **`INS_NOT_SUPPORTED` path (Nano S)** — existing behaviour preserved: device prompts with hashes only.
6. **Regression** — small single-message typed data on a blind-signing-capable device still clear-signs with one prompt; verify check passes silently.
7. **On-chain signature validity** — the broadcast signature's recovered signer matches the expected address at the dapp/RPC layer.

### **Before**

UI surfaces "Blind signing is not enabled" on any large EIP-712 payload, blocking the signature flow even with blind signing enabled on-device.

### **After**

Device prompts with domain/message hashes (same UX as the existing Nano S fallback); user confirms once; dapp receives a valid signature.

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I've included tests if applicable
- [x] I've documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I've applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

---

> [!NOTE]
> **Medium Risk**
> Changes hardware-wallet signing and fallback/error classification for EIP-712; incorrect handling could affect signature validity or user rejection UX, but scope is limited to the Ledger offscreen path with strong test coverage.
> 
> **Overview**
> Fixes Ledger **EIP-712** (`eth_signTypedData_v4`) signing when clear signing fails on large or complex payloads, which often surfaced as **"Blind signing is not enabled"** even with blind signing on.
> 
> In `ledger.ts`, the offscreen signer still tries `signEIP712Message` first, then **`signEIP712HashedMessage`**. The hashed path now runs when clear sign is rejected with **`INS_NOT_SUPPORTED`**, **`0x6808`**, or **`0x6a80`** (including codes only in the error message), not only `INS_NOT_SUPPORTED`. **`0x6985`** (on-device reject) still does **not** trigger a silent retry.
> 
> After a successful clear-sign response, the code **recovers the signer** (`recoverTypedSignature`) against the HD-path address; mismatch or malformed `r`/`s` (with normalization for prefixed/unpadded components) also falls through to hashed signing. If the hashed signature still fails verification, it throws a clear **`HardwareWalletError`**.
> 
> **CHANGELOG** and expanded **`ledger.test.ts`** cover happy path, fallbacks, user rejection, and verification edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 54cb6d78eaf5a5dec95a4382aecaa061829b48a6. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by ivan-angjelkoski** [comment]
I have read the CLA Document and I hereby sign the CLA

File: app/offscreen/hardware-wallets/ledger.ts [hunk 1]
```diff
@@ -360,12 +400,66 @@ export class LedgerOffscreenHandler {
...
+    }
+    const CLEAR_SIGN_REJECTION_CODES = new Set([0x6808, 0x6a80, 0x6985]);
+    if (
+      typeof statusCode === 'number' &&
+      CLEAR_SIGN_REJECTION_CODES.has(statusCode)
+    ) {
+      return true;
+    }
+    // Some transports surface the status code only in the message string.
+    return /0x(?:6808|6a80|6985)\b/iu.test(error.message);
```
[the remark this answers is not in the document]
Comment (ivan-angjelkoski) [reply app/offscreen/hardware-wallets/ledger.ts]: Resolved https://github.com/MetaMask/metamask-extension/pull/42010/changes/0fa511c07e9a5f3ce8ea521af9a3f7a59813148c

**Comment by ivan-angjelkoski** [comment]
Here is a demo app to test large payloads.
[https://ledger-eip712-payload-test-app.vercel.app/](https://ledger-eip712-payload-test-app.vercel.app/)

1. Connect Metamask
2. Press Auto (This creates the messages, the EIP712 data, and prompts the user to sign)

or if you want to manually trigger the steps:

1. Connect Metamask
2. Click `Build Messages`
3. Click `Build EIP712 Typed Data`
4. Click `Sign with Metamask`

in the `Signature` section you can see the details once you sign.

**Comment by owencraston** [comment]
@ivan-angjelkoski can you resolve the conflicts and post a video of this working to show off the user experience? Thanks in advance.

**Comment by ivan-angjelkoski** [comment]
Hey @owencraston , i recorded a loom video with the explanation and demo with and without the fix.
https://www.loom.com/share/62da549dfebd453999f68fe5b1d2edd8

**Comment by owencraston** [comment]
@ivan-angjelkoski I am having trouble testing this feature. I keep running into the following error (on main). My colleague @gantunesr experienced the same thing. 

Screenshot 2026-05-05 at 4 21 25 PM


We recently made some changes that would fall back to blind signing when clearSigning was not supported. I want to verify if our existing change will solve your problems. [This change was merged into main last week](https://github.com/MetaMask/metamask-extension/pull/41948) so you can follow the same steps you mentioned [here](https://github.com/MetaMask/metamask-extension/pull/42010#issuecomment-4319511266) with main and see if this unblocks you. If not, we'll take another look at your PR. Thanks in advance.

**Comment by ivan-angjelkoski** [comment]
Hey @owencraston , i fixed the issue on the test page, the account that you tried with didn't had an active account on injective, thats why it failed, so i changed it to use a test account for the payload instead and now it should work.

and the change that was already merged did not resolve this issue.

File: CHANGELOG.md [hunk 2]
```diff
@@ -296,6 +296,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Fixed a crash when selecting an external recipient during cross-chain bridge transactions (#41442)
 - Fixed order and funding transactions incorrectly appearing in perps Recent Activity (#41446)
 - Fixed cancel-all and close-all controls on perps (#41430)
+### Fixed
+
+- Fixed EIP-712 typed data signing on Ledger devices for large payloads, which previously failed with a misleading "Blind signing is not enabled" error even when blind signing was enabled; the signer now retries via hashed signing when clear signing is rejected, and verifies the returned signature against the expected address (#42010)
```
Comment (owencraston) [inline CHANGELOG.md:299]: You added a another fixed header here, can you just append it to the existing Fixed section?

File: app/offscreen/hardware-wallets/ledger.ts [hunk 3]
```diff
@@ -382,12 +424,71 @@ export class LedgerOffscreenHandler {
...
+      return false;
+    }
+    const { statusText, statusCode } = error as {
+      statusText?: string;
+      statusCode?: number;
+    };
+    if (statusText === 'INS_NOT_SUPPORTED') {
+      return true;
+    }
+    const CLEAR_SIGN_REJECTION_CODES = new Set([0x6808, 0x6a80]);
```
Comment (owencraston) [inline app/offscreen/hardware-wallets/ledger.ts]: This set gets created every invocation. I think it would make more sense to put this as a static read only property on the class.

File: app/offscreen/hardware-wallets/ledger.ts [hunk 4]
```diff
@@ -382,12 +424,71 @@ export class LedgerOffscreenHandler {
...
+          sig.s.replace(/^0x/u, '').padStart(64, '0') +
+          sig.v.toString(16).padStart(2, '0'),
+      );
+      const recovered = recoverTypedSignature({
+        data: message as TypedMessage<MessageTypes>,
+        signature,
+        version: SignTypedDataVersion.V4,
+      });
+      return recovered.toLowerCase() === expectedAddress.toLowerCase();
+    } catch {
```
Comment (owencraston) [inline app/offscreen/hardware-wallets/ledger.ts]: lets add some error logging here with the error message being logged.

File: app/offscreen/hardware-wallets/ledger.ts [hunk 5]
```diff
@@ -337,23 +352,50 @@ export class LedgerOffscreenHandler {
     params: LedgerSignTypedDataParams,
   ): Promise<{ v: number; r: string; s: string }> {
     const app = await this.ensureApp();
+    const { address: expectedAddress } = await app.getAddress(
```
Comment (owencraston) [inline app/offscreen/hardware-wallets/ledger.ts]: I dont like how we are adding an extra call to Ledger for cases where this fallback is not needed. could we move this inside the verifyTypedDataSignature or pass a lazy callback.

Comment (ivan-angjelkoski) [reply app/offscreen/hardware-wallets/ledger.ts]: I kept the address verification because it is the guard against clear signing returning a signature for a truncated/wrong payload, but changed the `getAddress` call to a lazy cached callback.

**Comment by ivan-angjelkoski** [comment]
Hey @owencraston, i resolved the comments. can you take another look?

**Comment by owencraston** [comment]
@ivan-angjelkoski I am having one of the engineers take a look soon. Give us a week or two for this. cc @gantunesr

File: app/offscreen/hardware-wallets/ledger.ts [hunk 6]
```diff
@@ -382,12 +443,76 @@ export class LedgerOffscreenHandler {
...
+  // message.
+  async #verifyTypedDataSignature(
+    message: LedgerSignTypedDataParams['message'],
+    sig: { v: number; r: string; s: string },
+    getExpectedAddress: () => Promise<string>,
+  ): Promise<boolean> {
+    let recovered: string;
+    try {
+      const signature = add0x(
+        sig.r.replace(/^0x/u, '').padStart(64, '0') +
```
Comment (owencraston) [inline app/offscreen/hardware-wallets/ledger.ts]: this regex concerns me for no other reason than I am suspicious of regex, is there a better/more reliable way we can achieve this? If not its okay, lets just make sure this has extensive testing. Perhaps we can pull this into another util function.

File: app/offscreen/hardware-wallets/ledger.ts [hunk 7]
```diff
@@ -382,12 +459,76 @@ export class LedgerOffscreenHandler {
...
+  // payload and that we should retry via hashed signing.
+  //
+  // - INS_NOT_SUPPORTED: older firmware (e.g. Nano S) that has no EIP-712
+  //   clear-signing support at all.
+  // - 0x6808 / 0x6a80: newer firmware rejecting an oversized or unparseable
+  //   payload. These bubble up to the UI as "Blind signing is not enabled"
+  //   via LEDGER_ERROR_MAPPINGS, even when blind signing IS enabled on the
+  //   device.
+  //
+  // 0x6985 (CONDITIONS_OF_USE_NOT_SATISFIED) is deliberately excluded: it is
```
Comment (gantunesr) [inline app/offscreen/hardware-wallets/ledger.ts:509]: nit: the PR description does say that the `0x6985` code is included in the fallback

File: app/offscreen/hardware-wallets/ledger.ts [hunk 8]
```diff
@@ -338,22 +373,64 @@ export class LedgerOffscreenHandler {
   ): Promise<{ v: number; r: string; s: string }> {
     const app = await this.ensureApp();
 
+    let expectedAddress: string | undefined;
+    const getExpectedAddress = async (): Promise<string> => {
+      if (expectedAddress === undefined) {
+        const { address } = await app.getAddress(params.hdPath, false);
+        expectedAddress = address;
+      }
+      return expectedAddress;
```
Comment (gantunesr) [inline app/offscreen/hardware-wallets/ledger.ts:382]: why do we get the address from the device and not just use the one that is already in the wallet? We could add the address to `LedgerSignTypedDataParams`

**Comment by owencraston** [comment]
Thanks for the thorough writeup @ivan-angjelkoski — the diagnosis is clear and the verify-after-sign step is a genuinely nice addition beyond Ledger's reference. I reviewed this with Daniel Rocha and Gustavo on
  our side. Before we merge, there's one question we need to answer first, because it determines whether this is the right fix at all:

  1. Is 0x6808 here actually "the device can't clear-sign this," or is it a payload we're malforming?

  The PR assumes 0x6808/0x6a80 means the payload is legitimately too complex for the device, so falling back to hashed signing is the correct response. But Ledger's Ethereum app is expected to clear-sign reasonably
  large and complex typed data — so it's not obvious these payloads are genuinely beyond it. If they're not, then 0x6808 is pointing at something else (e.g. how we're serializing the payload before handing it to
  the device), and a silent fallback would mask that root cause rather than fix it.

  So the key step is to craft a few representative payloads — large stringified contract-call JSON, multi-message Cosmos EIP-712, deeply nested/numerous domain types — and run them on real hardware (Nano X / S Plus
  / Stax, blind signing on) to establish: which of these does the device clear-sign fine, and which genuinely produce 0x6808? That tells us whether the fix belongs here (device truly can't), upstream at Ledger /
  ERC-7730 (firmware limitation we should push back on them), or in our own encoding path (our bug). The Loom shows the fallback working, but it doesn't answer whether the device should have clear-signed in the
  first place. Happy for us to take this on our end this week if that's easier.

  2. If it is a real device limitation, the fallback shouldn't be silent.

  Assuming the hardware test confirms the device genuinely can't clear-sign these, there's a second issue: this case isn't equivalent to the existing Nano S path. INS_NOT_SUPPORTED means the device can't clear-sign
  EIP-712 at all, so falling back loses nothing. 0x6808 means it normally can, just not this payload — so silently retrying with hashes moves the user from clear-signing to blind-signing without their knowledge.
  In that case we'd want a visible warning when the fallback fires ("Your device couldn't display this transaction in full — you'll be asked to approve hashes instead"), rather than a silent retry. The existing
  Nano S path staying silent is fine.

**Comment by owencraston** [comment]
@ivan-angjelkoski I have an update. We have just completed the [Ledger DMK migration](https://developers.ledger.com/docs/device-interaction/dmk-ts/integration/how_to/dmk) which should alleviate a lot of the issues you are having. The DMK is not currently in prod and will likely take a couple of weeks to release. Could you retest with a local build and verify that your issues have been fixed? You must use a local build since its not enabled in prod.
